lnd_test.go

```diff
@@ -13011,6 +13013,194 @@ func testSweepAllCoins(net *lntest.NetworkHarness, t *harnessTest) {
+	carol, err := net.NewNode("carol", []string{carolArgs})
+	if err != nil {
+		t.Fatalf("unable to create new node: %v", err)
+	}
+	defer shutdownAndAssert(net, t, carol)
+
+	// Next, we'll register for streaming notifications for changes to the
+	// backup file.
+	backupStream, err := carol.SubscribeChannelBackups(
+		ctxb, &lnrpc.ChannelBackupSubscription{},
+	)
+	if err != nil {
+		t.Fatalf("unable to create backup stream: %v", err)
+	}
+
+	// We'll use this goroutine to proxy any updates to a channel we can
+	// easily use below.
+	var wg sync.WaitGroup
+	backupUpdates := make(chan *lnrpc.ChanBackupSnapshot)
+	streamErr := make(chan error)
+	streamQuit := make(chan struct{})
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			snapshot, err := backupStream.Recv()
+			if err != nil {
+				select {
+				case streamErr <- err:
+				case <-streamQuit:
+					return
+				}
+			}
+
+			select {
+			case backupUpdates <- snapshot:
+			case <-streamQuit:
+				return
+			}
+		}
+	}()
+	defer close(streamQuit)
+
+	// With Carol up, we'll now connect her to Alice, and open a channel
+	// between them.
+	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+	if err := net.ConnectNodes(ctxt, carol, net.Alice); err != nil {
+		t.Fatalf("unable to connect carol to alice: %v", err)
+	}
+
+	// Next, we'll open two channels between Alice and Carol back to back.
+	var chanPoints []*lnrpc.ChannelPoint
+	numChans := 2
+	chanAmt := btcutil.Amount(1000000)
+	for i := 0; i < numChans; i++ {
+		ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+		chanPoint := openChannelAndAssert(
+			ctxt, t, net, net.Alice, carol,
+			lntest.OpenChannelParams{
+				Amt: chanAmt,
+			},
+		)
+
+		chanPoints = append(chanPoints, chanPoint)
+	}
+
+	// Using this helper function, we'll maintain a pointe rot he latest
+	// channel backup so we can compare it to the on disk state.
+	var currentBackup *lnrpc.ChanBackupSnapshot
+	assertBackupNtfns := func(numNtfns int) {
+		for i := 0; i < numNtfns; i++ {
+			select {
+			case err := <-streamErr:
+				t.Fatalf("error with backup stream: %v", err)
+
+			case currentBackup = <-backupUpdates:
+
+			case <-time.After(time.Second * 5):
+				t.Fatalf("didn't receive channel backup "+
+					"notification %v", i+1)
+			}
+		}
+	}
+
+	// assertBackupFileState is a helper function that we'll use to compare
+	// the on disk back up file to our currentBackup pointer above.
+	assertBackupFileState := func() {
+		err := lntest.WaitNoError(func() error {
+			packedBackup, err := ioutil.ReadFile(backupFilePath)
+			if err != nil {
+				return fmt.Errorf("unable to read backup "+
+					"file: %v", err)
+			}
+
+			// As each back up file will be encrypted with a fresh
+			// nonce, we can't compare them directly, so instead
+			// we'll compare the length which is a proxy for the
+			// number of channels that the multi-backup contains.
+			rawBackup := currentBackup.MultiChanBackup.MultiChanBackup
+			if len(rawBackup) != len(packedBackup) {
```

Comment:
can we compare the bytes directly?

Reply:
No, since we use a random nonce each time to create a backup. I created a the `VerifyChanBackup` command to help with this, as it allows users to ensure they have a valid backup, since it can change on disk.